beesdoo_product/models/beesdoo_product.py

```diff
@@ -363,6 +333,91 @@ def _inverse_purchase_price(self):
                     _("No Vendor defined for product '%s'") % product.name
                 )
 
+    @api.multi
+    @api.depends("purchase_price", "seller_ids")
+    def _compute_purchase_price_write_date(self):
+        for product in self:
+            supplierinfo = product._get_main_supplier_info()
+            product.purchase_price_write_date = supplierinfo.price_write_date
+
+    def write(self, vals):
+        purchase_price = vals.get("purchase_price")
+        if purchase_price and purchase_price != self.purchase_price:
+            try:
+                # We could write suggested_price to vals _now_, but the compute
+                # function is going to run regardless. It's a bit useless to run
+                # 'calculate_suggested_price' twice, but that's just how it is.
+                suggested_price = self.calculate_suggested_price(price=purchase_price)
+
+                # Important note: The list price is _only_ changed here if the
+                # `purchase_price` field of the product is changed. If the
+                # `price` field of the supplierinfo changes, the list price here
+                # is NOT automatically affected.
+                self.adapt_list_price(vals, suggested_price=suggested_price)
+            except ValueError:
+                pass
```

Comment:
the rationale behind `adapt_list_price()` was to avoid duplicating the logic (that determines where the new price is coming from) in the wizard and in `write()`. as it is, the function is not useful anymore. the problem comes from the fact that the wizard can use `product.suggested_price`, while in `write()` it’s not possible because the property is not yet up to date. could we maybe have something like this?
```suggestion
            # force update of purchase_price (which actually modifies the
            # suppliers’ price) to ensure that the next computations that
            # depend on it are correct. there are 3 things to do:
            # 1. set value in cache
            with self.env.do_in_draft():
                self.purchase_price = purchase_price
            # 2. call inverse compute method
            self._inverse_purchase_price()
            # 3. remove the value from vals, to avoid the process to happen a
            #    second time
            del vals["purchase_price"]

            self.adapt_list_price(vals)
```

this allows to keep the code as it was before: no need to move `suggested_price` computation in another method, and the relationship between `list_price` and `suggested_price` is only in `adapt_list_price()`. the downside is that it’s probably not very clean to modify fields like this, and that `_inverse_purchase_price()` is called explicitly in `write()` instead of using a generic method (but maybe there is a way?).

Reply:
I'm not entirely sure this works, and if it does work, I think it might break a (possibly implicit) rule:

1. `self.purchase_price` is set in draft mode and written to a cache. How long does this cache live? When is that cache emptied? Fortunately it's not a stored field, but if it were, this would assuredly break.
2. `self._inverse_purchase_price()` writes before `write()` is done. I'm not sure that sort of non-atomic operation is allowed. If `write()` is in any way interrupted, you get an inconsistent cache/database.
3. If `self._inverse_purchase_price()` writes something to the main supplier, then isn't `self.seller_ids` changed, and isn't `self._compute_cost()` going to trigger?

This is a stupid problem :/

Reply:
Tested this and it works. Haven't found a way to break it. I'll push this as commit. If it doesn't work as desired, we can revert that specific commit.